src/main/java/net/schmizz/sshj/connection/channel/Window.java

```diff
@@ -73,17 +75,23 @@ public String toString() {
     /** Controls how much data we can send before an adjustment notification from remote end is required. */
     public static final class Remote
             extends Window {
+        private final long timeoutMs;
 
-        public Remote(long initialWinSize, int maxPacketSize, LoggerFactory loggerFactory) {
+        public Remote(long initialWinSize, int maxPacketSize, long timeoutMs, LoggerFactory loggerFactory) {
             super(initialWinSize, maxPacketSize, loggerFactory);
+            this.timeoutMs = timeoutMs;
         }
 
         public long awaitExpansion(long was) throws ConnectionException {
             synchronized (lock) {
+                long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeoutMs);
```

Comment:
`System.currentTimeMillis` has enough granularity for a timeout ;). That also prevents an unnecessary conversion.

Reply:
It isn’t monotomic, so it may fluxuate (eg NTP sync), so typically it’s avoided for non datetime logic. Do you still prefer it?

Reply:
Happened again. I think this is correct, can you take a look?

Reply:
ok, agreed.